airflow/models.py

```diff
@@ -573,11 +573,14 @@ def __init__(
             self.extra = extra
 
     def parse_from_uri(self, uri):
-        temp_uri = urlparse(uri)
+        temp_scheme, rest = uri.split('//', 1)
+        # For a scheme can not be parsed (See https://issues.apache.org/jira/browse/AIRFLOW-1330)
+        temp_uri = '//'.join([temp_scheme.replace('_', '-'), rest])
```

Comment:
Can we leave parse_from_uri untouched?

Reply:
Tell me why. What is your concern?

Reply:
Given that we add conn_type, ..., we should expect parse_from_uri would work out-of-the-box without any hack. IMHO, I am thinking:
- either we fail silently like what this does today
- or we perform uri validation and error out early.  

What you have here would still allow someone to pass in --conn_uri=scheme_with_underscore and succeed, what if there are other illegal scheme characters, we then have to patch parse_from_uri from time to time (it's better to not open any loophole).

Reply:
Even though the RFC illegal character is included in the schema, if it is a valid conn_type of Airflow, I think that should not cause an error.
Of course, I think that it is ideal to match conn_type to RFC.

Reply:
It probably should cause an error on --conn-uri as it's not a valid URI. On the other hand, like you said, if it's a valid airflow conn-type, we should accept it where the user however needs to manually specify conn-type, hostname, password, etc.

Reply:
Umm... After all I only need to be able to create a conn_type `google_cloud_platform` connection from the CLI, so I do not mind excluding url_parse hack from this PR.
@criccomini How do you think?

Reply:
Just want to note that GCP takes in inputs like project id, keyfile, and scope rather than connection host, port, login, password, etc. So the conn_uri parsing doesn't apply.

There are currently 3 other incompatible connection types in Airflow: `mesos_framework-id`, `hive_cli` and `hive_metastore`. The mesos one only uses the `extra` field so conn_uri doesn't apply to it either. The hive connections needs host/port, so it would cause an issue if someone is trying to add connections with conn_uri. 

Personally I agree with @fenglu-g on aligning with the RFC standards, and only provide connections with proper URI to be parsed correctly to minimize loopholes. Going forward, we probably want to create guards to prevent adding connection types with `_`.

Reply:
Fair enough. I will change my PR.